contracts/booking-token/CancellationUtils.sol

```diff
@@ -0,0 +1,176 @@
+ */
+library CancellationUtils {
+    // Constants for bit manipulation
+    uint256 private constant SEGMENT_MASK = 0xFFFF; // 16 bits mask
+    uint256 private constant CANCELLATION_REASON_SHIFT = 0; // First 16 bits
+    uint256 private constant CANCELLATION_VERSION_SHIFT = 16; // Next 16 bits
+    uint256 private constant REJECTION_REASON_SHIFT = 32; // Next 16 bits
+    uint256 private constant REJECTION_VERSION_SHIFT = 48; // Final 16 bits
+
+    // Masks for different sections
+    uint256 private constant CANCELLATION_MASK = 0xFFFFFFFF;
+    uint256 private constant REJECTION_MASK = 0xFFFFFFFF00000000;
+
+    // Custom errors
+    error ValueExceedsMaxSize(uint256 value, uint256 maxSize);
+
+    /**
+     * @notice Packs cancellation and rejection data into a single uint256
+     * @param cancellationReason The reason for cancellation (max 16 bits)
+     * @param cancellationVersion Version of the cancellation reason enum (max 16 bits)
+     * @param rejectionReason The reason for rejection (max 16 bits)
+     * @param rejectionVersion Version of the rejection reason enum (max 16 bits)
+     * @return packed The packed uint256 containing all data
+     */
+    function packReasons(
+        uint256 cancellationReason,
+        uint256 cancellationVersion,
+        uint256 rejectionReason,
+        uint256 rejectionVersion
+    ) internal pure returns (uint256) {
+        // Check if any value exceeds 16 bits
+        if (cancellationReason > SEGMENT_MASK) {
+            revert ValueExceedsMaxSize(cancellationReason, SEGMENT_MASK);
+        }
+        if (cancellationVersion > SEGMENT_MASK) {
+            revert ValueExceedsMaxSize(cancellationVersion, SEGMENT_MASK);
+        }
+        if (rejectionReason > SEGMENT_MASK) {
+            revert ValueExceedsMaxSize(rejectionReason, SEGMENT_MASK);
+        }
+        if (rejectionVersion > SEGMENT_MASK) {
+            revert ValueExceedsMaxSize(rejectionVersion, SEGMENT_MASK);
+        }
+
+        // Pack all values into a single uint256
+        return
+            (rejectionVersion << REJECTION_VERSION_SHIFT) |
+            (rejectionReason << REJECTION_REASON_SHIFT) |
+            (cancellationVersion << CANCELLATION_VERSION_SHIFT) |
+            (cancellationReason << CANCELLATION_REASON_SHIFT);
+    }
+
+    /**
+     * @notice Unpacks a uint256 into its component values
+     * @param packed The packed uint256 containing all data
+     * @return cancellationReason The reason for cancellation
+     * @return cancellationVersion Version of the cancellation reason enum
+     * @return rejectionReason The reason for rejection
+     * @return rejectionVersion Version of the rejection reason enum
+     */
+    function unpackReasons(
+        uint256 packed
+    )
+        internal
+        pure
+        returns (
+            uint256 cancellationReason,
+            uint256 cancellationVersion,
+            uint256 rejectionReason,
+            uint256 rejectionVersion
```

Comment:
Why return uint256 if the values are defined as uint16?

Reply:
Because I decided to use `uint256` for the pack function, I kept it the same also here.